backend/plugin/advisor/mysqlwip/advisor_column_disallow_changing_type.go

```diff
@@ -57,85 +64,100 @@ func (*ColumnDisallowChangingTypeAdvisor) Check(ctx advisor.Context, _ string) (
 }
 
 type columnDisallowChangingTypeChecker struct {
+	*mysql.BaseMySQLParserListener
+
+	baseLine   int
 	adviceList []advisor.Advice
 	level      advisor.Status
 	title      string
 	text       string
-	line       int
 	catalog    *catalog.Finder
 }
 
-// Enter implements the ast.Visitor interface.
-func (checker *columnDisallowChangingTypeChecker) Enter(in ast.Node) (ast.Node, bool) {
-	changeType := false
-	if node, ok := in.(*ast.AlterTableStmt); ok {
-		for _, spec := range node.Specs {
-			switch spec.Tp {
-			case ast.AlterTableChangeColumn:
-				changeType = checker.changeColumnType(node.Table.Name.O, spec.OldColumnName.Name.O, spec.NewColumns[0].Tp.String())
-			case ast.AlterTableModifyColumn:
-				changeType = checker.changeColumnType(node.Table.Name.O, spec.NewColumns[0].Name.Name.O, spec.NewColumns[0].Tp.String())
-			}
-			if changeType {
-				break
-			}
-		}
-	}
+func (checker *columnDisallowChangingTypeChecker) EnterQuery(ctx *mysql.QueryContext) {
+	checker.text = ctx.GetParser().GetTokenStream().GetTextFromRuleContext(ctx)
+}
 
-	if changeType {
-		checker.adviceList = append(checker.adviceList, advisor.Advice{
-			Status:  checker.level,
-			Code:    advisor.ChangeColumnType,
-			Title:   checker.title,
-			Content: fmt.Sprintf("\"%s\" changes column type", checker.text),
-			Line:    checker.line,
-		})
+// EnterAlterTable is called when production alterTable is entered.
+func (checker *columnDisallowChangingTypeChecker) EnterAlterTable(ctx *mysql.AlterTableContext) {
+	if ctx.AlterTableActions() == nil {
+		return
+	}
+	if ctx.AlterTableActions().AlterCommandList() == nil {
+		return
+	}
+	if ctx.AlterTableActions().AlterCommandList().AlterList() == nil {
+		return
 	}
 
-	return in, false
+	_, tableName := mysqlparser.NormalizeMySQLTableRef(ctx.TableRef())
+	// alter table add column, change column, modify column.
+	for _, item := range ctx.AlterTableActions().AlterCommandList().AlterList().AllAlterListItem() {
+		if item == nil {
+			continue
+		}
+
+		var columnName string
+		switch {
+		// change column
+		case item.CHANGE_SYMBOL() != nil && item.ColumnInternalRef() != nil && item.Identifier() != nil:
+			// only focus on old colunn-name.
+			columnName = mysqlparser.NormalizeMySQLColumnInternalRef(item.ColumnInternalRef())
+		// MODIFY COLUMN
+		case item.MODIFY_SYMBOL() != nil && item.ColumnInternalRef() != nil && item.FieldDefinition() != nil:
+			columnName = mysqlparser.NormalizeMySQLColumnInternalRef(item.ColumnInternalRef())
+		default:
+			continue
+		}
+		checker.changeColumnType(tableName, columnName, item.FieldDefinition().DataType())
+	}
 }
 
-// Leave implements the ast.Visitor interface.
-func (*columnDisallowChangingTypeChecker) Leave(in ast.Node) (ast.Node, bool) {
-	return in, true
+func normalizeColumnTypeImpl(tp string) string {
+	switch {
+	case strings.HasPrefix(tp, "integer"), strings.HasPrefix(tp, "int"):
```

Comment:
consider that `int(1)` and `int(2)`

Reply:
> MySQL supports an extension for optionally specifying the display width of integer data types in parentheses following the base keyword for the type.

and this

> The display width does not constrain the range of values that can be stored in the column. Nor does it prevent values wider than the column display width from being displayed correctly.

https://dev.mysql.com/doc/refman/8.0/en/numeric-type-attributes.html